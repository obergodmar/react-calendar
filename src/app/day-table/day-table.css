.row {
    display: flex;
    justify-content: space-between;
    color: black;
    text-align: center;
}

.list,
.row {
    list-style: none;
}

.weekday {
    width: 28px;
    height: 20px;
}

.day {
    position: relative;
    z-index: 1;
    display: flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    width: 28px;
    height: 28px;
    border: 1px solid transparent;
    border-radius: 2px;
    cursor: default;
}

.day:focus {
    outline: none;
}

.day:focus,
.day:hover {
    background-color: #b2d5ff;
}

.dayOther {
    color: gray;
}

.dayToday {
    border: 1px solid black;
}

.daySelected {
    color: #fff;
    font-weight: 600;
    background-color: #0075ff;
    border: 1px solid black;
    border-radius: 0;
}

.daySelected:focus,
.daySelected:hover {
    background-color: #0075ff;
    border-radius: 0;
}
