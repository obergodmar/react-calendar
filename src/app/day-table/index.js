export { DayTable } from "./day-table";
