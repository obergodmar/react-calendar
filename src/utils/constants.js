export const TEST_DATE = new Date(1970, 1, 1);
export const FULL_FORMAT = "MMddyyyy";
export const WEEKDAY_FORMAT = "EEEEEE";
export const MONTH_FORMAT = "MMM";
export const MONTH_YEAR = "LLLL yyyy";
export const DATE_FORMAT = "MM / dd / yyyy";
