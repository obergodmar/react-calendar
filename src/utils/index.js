export { default as dayMatrix } from "./day-matrix";
export * from "./constants";
