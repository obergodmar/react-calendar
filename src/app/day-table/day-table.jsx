import React, { Fragment, useCallback, useMemo } from "react";
import PropTypes from "prop-types";
import cn from "classnames";
import format from "date-fns/format";
import addDays from "date-fns/addDays";
import startOfWeek from "date-fns/startOfWeek";

import { TEST_DATE, WEEKDAY_FORMAT } from "../../utils";

import style from "./day-table.css";

export const DayTable = ({ calendar, changeDate }) => {
  // TODO: Add locale
  const weekdays = useMemo(() => {
    let date = startOfWeek(TEST_DATE);
    return [...Array(7).keys()].map(() => {
      const weekday = format(date, WEEKDAY_FORMAT);
      date = addDays(date, 1);
      return weekday;
    });
  }, []);

  const handlePreventDefault = useCallback(
    (event) => event.preventDefault(),
    []
  );

  const handleKeyDown = useCallback(
    (event, date, needClose) => {
      const { keyCode } = event;
      if (keyCode === 13 || keyCode === 32) {
        changeDate(date, needClose);
      }
    },
    [changeDate]
  );

  return (
    <Fragment>
      <ul className={style.row}>
        {weekdays.map((day, idx) => (
          <li key={idx} className={style.weekday}>
            {day}
          </li>
        ))}
      </ul>
      <ul className={style.list}>
        {calendar.map((row, rowIdx) => (
          <li key={rowIdx}>
            <ul className={style.row}>
              {row.map((day, dayIdx) => (
                <li
                  tabIndex={0}
                  onMouseDown={handlePreventDefault}
                  onClick={() => changeDate(day.date, day.isPart)}
                  onKeyDown={(e) => handleKeyDown(e, day.date, day.isPart)}
                  key={dayIdx}
                  className={cn(style.day, {
                    [style.dayOther]: !day.isPart,
                    [style.dayToday]: day.isToday,
                    [style.daySelected]: day.isSelected,
                  })}
                >
                  {day.date.getDate()}
                </li>
              ))}
            </ul>
          </li>
        ))}
      </ul>
    </Fragment>
  );
};

DayTable.propTypes = {
  calendar: PropTypes.arrayOf(
    PropTypes.arrayOf(
      PropTypes.shape({
        date: PropTypes.instanceOf(Date),
        isToday: PropTypes.bool,
        isSelected: PropTypes.bool,
        isPart: PropTypes.bool,
      })
    )
  ).isRequired,
  changeDate: PropTypes.func.isRequired,
};

DayTable.displayName = "DayTable";
