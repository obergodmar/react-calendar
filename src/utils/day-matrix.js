import { addDays, format, startOfWeek } from "date-fns";
import { FULL_FORMAT } from "./constants";

const rows = 6;
const cols = 7;
const length = rows * cols;

export default (currentDate) => {
  const today = format(new Date(), FULL_FORMAT);
  const currentDateFormat = format(currentDate, FULL_FORMAT);
  const year = currentDate.getFullYear();
  const month = currentDate.getMonth();
  const startDate = startOfWeek(new Date(year, month, 1));
  return Array.from({ length })
    .map((_, index) => {
      const date = addDays(startDate, index);
      const dateFormat = format(date, FULL_FORMAT);
      const currentMonth = date.getMonth();
      const isPart = currentMonth === month;
      const isToday = dateFormat === today;
      const isSelected = currentDateFormat === dateFormat;
      return {
        date,
        isToday,
        isSelected,
        isPart,
      };
    })
    .reduce(
      (matrix, _, index, days) =>
        !(index % cols !== 0)
          ? [...matrix, days.slice(index, index + cols)]
          : matrix,
      []
    );
};
