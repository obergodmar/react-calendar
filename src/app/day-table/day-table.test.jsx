import React from "react";
import { describe, jest, test } from "@jest/globals";
import { shallow } from "enzyme";

import { dayMatrix, TEST_DATE } from "../../utils";
import { DayTable } from "./day-table";

describe("Day Table", () => {
  const handleChangeDate = jest.fn();
  const calendar = dayMatrix(new Date(TEST_DATE));
  const wrapper = shallow(
    <DayTable calendar={calendar} changeDate={handleChangeDate} />
  );

  test("Should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Should change date by click", () => {
    wrapper.find(".day").at(2).simulate("click");
    expect(handleChangeDate).toHaveBeenCalled();
  });

  test("Should change date by keydown", () => {
    wrapper.find(".day").at(5).simulate("keydown", { keyCode: 13 });
    expect(handleChangeDate).toHaveBeenCalled();

    wrapper.find(".day").at(7).simulate("keydown", { keyCode: 32 });
    expect(handleChangeDate).toHaveBeenCalled();
  });

  test("Should not change date by keydown", () => {
    const newChangeMonth = jest.fn();
    const component = shallow(
      <DayTable calendar={calendar} changeDate={handleChangeDate} />
    );

    component.find(".day").at(5).simulate("keydown", { keyCode: 37 });
    expect(newChangeMonth).not.toHaveBeenCalled();
  });

  test("Should be default prevented", () => {
    const event = { preventDefault: jest.fn() };

    wrapper.find(".day").at(10).simulate("mousedown", event);
    expect(event.preventDefault).toHaveBeenCalledTimes(1);
  });

  test("Should be 1st of January", () => {
    expect(wrapper.find(".day").at(0).is(".daySelected")).toEqual(true);
  });
});
