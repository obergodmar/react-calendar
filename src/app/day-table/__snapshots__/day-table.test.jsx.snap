// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day Table Should match snapshot 1`] = `
<Fragment>
  <ul
    className="row"
  >
    <li
      className="weekday"
      key="0"
    >
      Su
    </li>
    <li
      className="weekday"
      key="1"
    >
      Mo
    </li>
    <li
      className="weekday"
      key="2"
    >
      Tu
    </li>
    <li
      className="weekday"
      key="3"
    >
      We
    </li>
    <li
      className="weekday"
      key="4"
    >
      Th
    </li>
    <li
      className="weekday"
      key="5"
    >
      Fr
    </li>
    <li
      className="weekday"
      key="6"
    >
      Sa
    </li>
  </ul>
  <ul
    className="list"
  >
    <li
      key="0"
    >
      <ul
        className="row"
      >
        <li
          className="day daySelected"
          key="0"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          1
        </li>
        <li
          className="day"
          key="1"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          2
        </li>
        <li
          className="day"
          key="2"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          3
        </li>
        <li
          className="day"
          key="3"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          4
        </li>
        <li
          className="day"
          key="4"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          5
        </li>
        <li
          className="day"
          key="5"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          6
        </li>
        <li
          className="day"
          key="6"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          7
        </li>
      </ul>
    </li>
    <li
      key="1"
    >
      <ul
        className="row"
      >
        <li
          className="day"
          key="0"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          8
        </li>
        <li
          className="day"
          key="1"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          9
        </li>
        <li
          className="day"
          key="2"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          10
        </li>
        <li
          className="day"
          key="3"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          11
        </li>
        <li
          className="day"
          key="4"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          12
        </li>
        <li
          className="day"
          key="5"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          13
        </li>
        <li
          className="day"
          key="6"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          14
        </li>
      </ul>
    </li>
    <li
      key="2"
    >
      <ul
        className="row"
      >
        <li
          className="day"
          key="0"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          15
        </li>
        <li
          className="day"
          key="1"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          16
        </li>
        <li
          className="day"
          key="2"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          17
        </li>
        <li
          className="day"
          key="3"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          18
        </li>
        <li
          className="day"
          key="4"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          19
        </li>
        <li
          className="day"
          key="5"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          20
        </li>
        <li
          className="day"
          key="6"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          21
        </li>
      </ul>
    </li>
    <li
      key="3"
    >
      <ul
        className="row"
      >
        <li
          className="day"
          key="0"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          22
        </li>
        <li
          className="day"
          key="1"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          23
        </li>
        <li
          className="day"
          key="2"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          24
        </li>
        <li
          className="day"
          key="3"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          25
        </li>
        <li
          className="day"
          key="4"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          26
        </li>
        <li
          className="day"
          key="5"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          27
        </li>
        <li
          className="day"
          key="6"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          28
        </li>
      </ul>
    </li>
    <li
      key="4"
    >
      <ul
        className="row"
      >
        <li
          className="day dayOther"
          key="0"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          1
        </li>
        <li
          className="day dayOther"
          key="1"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          2
        </li>
        <li
          className="day dayOther"
          key="2"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          3
        </li>
        <li
          className="day dayOther"
          key="3"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          4
        </li>
        <li
          className="day dayOther"
          key="4"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          5
        </li>
        <li
          className="day dayOther"
          key="5"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          6
        </li>
        <li
          className="day dayOther"
          key="6"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          7
        </li>
      </ul>
    </li>
    <li
      key="5"
    >
      <ul
        className="row"
      >
        <li
          className="day dayOther"
          key="0"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          8
        </li>
        <li
          className="day dayOther"
          key="1"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          9
        </li>
        <li
          className="day dayOther"
          key="2"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          10
        </li>
        <li
          className="day dayOther"
          key="3"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          11
        </li>
        <li
          className="day dayOther"
          key="4"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          12
        </li>
        <li
          className="day dayOther"
          key="5"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          13
        </li>
        <li
          className="day dayOther"
          key="6"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          tabIndex={0}
        >
          14
        </li>
      </ul>
    </li>
  </ul>
</Fragment>
`;
